### ISSUE TITLE
terra-jupyter-r question about multi-core operations

### ISSUE THREAD

**Issue by sjfleming** [DataBiosphere/terra-docker#210]
Hello!  This is a question that's going to be a little bit incomplete... but here it goes:

I have been basing an image off of the `terra-jupyter-r:1.0.4` base image for use in Terra notebooks.

I just recently realized that, when I try to run commands like `lmFit()` in the `limma` package (for differential expression analysis), only one processor is being used.  When I have set up my own R installations in the past, `limma`'s differential expression testing has always automatically parallelized itself over all the available cores.  Unfortunately I do not understand much about multi-core computing in R... but I have come to rely on that factor of 16 speedup I get on a 16-cpu machine.

My question:
Is there something about the R installation in `terra-jupyter-r` that changes a default somewhere for how multi-core operations happen?  Something that might have broken the default behavior, which is to parallelize operations over all available cores?

(A bit more information... if I would watch processes in `top` in the past, there were two steps where I would see multi-core operations: one would show one process using 1600% compute, and the other operation would show 16 processes each using 100% compute.  I am guessing they used different means of parallelizing their operations.  But they both work on my own R installation, and they both fail to work (they show one process using 100% compute) on `terra-jupyter-r`.
Working install shows
```
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS

Matrix products: default
BLAS/LAPACK: /home/sfleming/miniconda3/envs/scanpy15/lib/libopenblasp-r0.3.7.so
```
`terra-jupyter-r:1.0.4` shows
```
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
```
)
I am not sure what else to look for!

Any tips would be much appreciated!

**Comment by sjfleming** [comment]
It looks like this could be addressed in a number of ways (either directly in R, or using conda), by installing OpenBLAS

**Comment by rtitle** [comment]
Hey @sjfleming!

Sorry I didn't see this until now. I am not sure the answer, but am curious. Do you have any example R code that demonstrates the issue?

**Comment by rtitle** [comment]
Yeah installing an optimized BLAS sounds like it could be a good idea: https://csantill.github.io/RPerformanceWBLAS/

**Comment by rtitle** [comment]
I simply ran in my runtime:
```
sudo apt-get install libopenblas-base
```
and I now see:
```
> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
```

Compared to previous:
```
> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
```

So @sjfleming if you add that `apt-get` line to your custom image, does it cause R to use all cores? If that fixes it, we can add that to the `terra-jupyter-r` image I think.

**Comment by rtitle** [comment]
Also (this is a long shot), we recently changed the default notebook runtime in the UI to a 1 CPU machine. So just double-check your runtime actually has >1 CPUs. :)

**Comment by nturaga** [comment]
You can try `parallel::detectCores()`

**Comment by sjfleming** [comment]
Hi @rtitle thanks for your response!  Yes, I think your `apt-get` solution might work, and I will let you know if it works when included in a custom docker image built on `terra-jupyter-r`.

Here's an example of a minimal piece of code that demonstrates the difference:

```R
m <- matrix(rnorm(9000000), nrow=3000)
d <- solve(m)
```

In my little experiments, this takes about 30 seconds with the current setup, but on a 16-cpu machine with OpenBLAS installed, it runs in 662ms.  If you watch `top` while the code runs, you can see that all the cores blast off when using OpenBLAS (though it's all finished very quickly anyway).

**Comment by sjfleming** [comment]
Image built on top of `terra-jupyter-r`
<img width="459" alt="image" src="https://user-images.githubusercontent.com/10214815/117075703-4475a580-ad03-11eb-8eee-454820d6f21f.png">

An install with OpenBLAS on a 16 CPU machine
<img width="444" alt="image" src="https://user-images.githubusercontent.com/10214815/117075868-8868aa80-ad03-11eb-97e1-2b10ecdcaa79.png">

**Comment by rtitle** [comment]
Cool, let us know how it goes @sjfleming. I also created a JIRA to track the work if we decide to do this on the Terra side: https://broadworkbench.atlassian.net/browse/IA-2736

**Comment by sjfleming** [comment]
Alright @rtitle ... I tried interactively running

```bash
sudo apt-get update
sudo apt-get install -yq --no-install-recommends libopenblas-base
```

inside my docker image (built on top of `terra-jupyter-r:1.0.4`).  This installation succeeded (if I ran as root), and when I ran R, it seemed like the above test passed.  So that's good news, and it seems promising.

But... I guess I don't have enough experience with this, but if I actually try to include the lines

```
USER root
RUN sudo apt-get update \
 && sudo apt-get install -yq --no-install-recommends libopenblas-base
USER $USER
```
in my own Dockerfile, then I always get the following error, which I can't yet find my way around:
```
W: GPG error: http://packages.cloud.google.com/apt cloud-sdk-bionic InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY FEEA9169307EA071 NO_PUBKEY 8B57C5C2836F4BEB
E: The repository 'http://packages.cloud.google.com/apt cloud-sdk-bionic InRelease' is not signed.
W: GPG error: https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease: The following signatures were invalid: EXPKEYSIG 51716619E084DAB9 Michael Rutter <marutter@gmail.com>
E: The repository 'https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease' is not signed.
```

(Please let me know if you happen to know the solution for this kind of error.)

**Comment by kyuksel** [comment]
@sjfleming I'm curious if including the line 

```
curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
```

or the line

```
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key --keyring /usr/share/keyrings/cloud.google.gpg add -
```

before the command

```sudo apt-get update```

would fix the issue.

**Comment by rtitle** [comment]
Maybe there was an edit.. sounds like @sjfleming is saying the above lines did the trick.

I agree we should fold that in to the `terra-jupyter-r` image, and perhaps also the Bioconductor RStudio [image](https://github.com/anvilproject/anvil-docker/tree/master/anvil-rstudio-bioconductor) (which is a separate Docker hierarchy).

**Comment by sjfleming** [comment]
Thanks @kyuksel , I realized I had screwed with the key I was using for R later on in my image... but if I put the necessary lines _before_ the part where I was messing around later, then it worked alright.  I'm grateful for the suggestion though if I run into that again, which I probably will.  I never know what to do in those cases, so that's very helpful.

@rtitle sounds great!
